active/0051-strategies.md

```diff
@@ -0,0 +1,368 @@
+
+#### Engine Strategy
+
+Introducing `Strategy` as a concept allows us to re-think the way we approach
+Ember Engines as well. We can start thinking about engines as a strategy of
+certain type. In fact, it's very similar to `ConcatenationStrategy` described
+above.
+
++ reduces the API surface (as opposed to `ember-engines` and
+  `ember-asset-loader`, only one strategy)
++ makes things explicit, meaning that you have to register a strategy
+
+## Registering custom strategies
+
+Ember applications would be able to register custom strategies via
+`ember-cli-build.js` file by passing a `strategies` property to the constructor.
+
+Passing `strategies` as an option to the constructor will clobber default
+strategies that Ember CLI provides out of the box (they mimic existing behaviour
+of creating `dist` folder with final assets).
+
+However, Ember CLI will expose an array of default strategies that people can
+take advantage of. Note, that `defaultStrategies` will be freezed so you can't
+push directly onto it. One would need to create a new array.
+
+For example:
+
+```javascript
+// ember-cli-build.js
+const { defaultStrategies } = require('ember-cli/lib/strategies');
+
+module.exports = function(defaults) {
+  const firstStrategy = new MyCustomStrategy();
+  const secondStrategy = new MyCustomStrategy();
+  const strategies = defaultStrategies.concat(firstStrategy, secondStrategy);
+
+  const app = new EmberApp(defaults, { strategies });
+
+  return app.toTree();
+}
+```
+
+Note, that `strategies` is an optional property. If you don't pass it to the
+constructor or `strategies` is an empty array, you're effectively "opting out"
+of the feature and using the default behaviour.
+
+This change should be behind an experiment flag, `STRATEGIES`. This will allow
+us to start experimenting with different strategies right away and not being
+tied to a particular release cycle.
+
+#  Topics for Future RFCs
+
+While working on this RFC, some ideas were brought into focus regarding existing
+and new features in Ember CLI. They all likely require separate discussions in
+future RFCs, but the discussion points have been included below.
+
+## Treeshaking
+
+Firstly, what's _tree-shaking_? AFAIK, the term
+[originated](https://groups.google.com/forum/#!msg/comp.lang.lisp/6zpZsWFFW18/-z_8hHRAIf4J)
+in Lisp. The gist of the idea is "how about we start using _only_ the code that
+we actually need?"
+
+Secondly, how is it different from [Dead Code
+Elimination](https://en.wikipedia.org/wiki/Dead_code_elimination)? [Rich
+Harris](https://twitter.com/Rich_Harris)
+[offers](https://medium.com/@Rich_Harris/tree-shaking-versus-dead-code-elimination-d3765df85c80)
+a pretty good explanation in the context of [Rollup](https://rollupjs.org/). The
+gist is dead code elimination happens on a final product by removing bits that
+are unused. Treeshaking is not about excluding the code but about including the
```

Comment:
"Treeshaking is not about excluding the code but about including the code that's used."
This line is super confusing to me and actually makes it harder for me to understand the difference. It could be just me though.